File: fushi/test/lookup/phrase_lookup_whitespace_bridge_bug1773_test.js
// BUG-1773 行为守卫：取词的前向扫描必须把空格当**词间连接符**，而不是终点。
//
// 症状：英文正文里点 "listen" 只查到 listen，`listen to` 这类空格分词短语的词条
// 永远匹配不到——因为 selectFromPosition 的前向扫描一遇空白就 break，喂给引擎的
// 查询串被截成单个单词。C++ scan_candidates 本来就按空格分词生成
// `listen to music` / `listen to` / `listen` 三级候选（禁止在单词中间切），拿不到
// 空格就等于把短语整类排除。
//
// 根因修复：把混成一坨的 isScanBoundary 拆成两个语义不同的谓词——
//   isScanBoundary（词边界：点击命中判定 + 词首回退用，含空白，语义不变）
//   isScanStop    （扫描终点：标点 / 门控，**不含空白**）
// 空白能否跨过去由桥接规则单独决定：只在**同一文本节点内部**跨、且只跨一个
// （左边必须已有本节点扫入的内容，右边必须紧跟一个可扫字符）。
//
// 本 harness 用 node:vm 在最小 fake DOM 里真执行两份实现的 selectFromPosition：
//   ① assets/popup/selection.js（浮窗 / 浏览器扩展，三镜像 parity 另有测试守）
//   ② 阅读器注入脚本 ReaderSelectionScripts.source()——由 .dart wrapper 落到临时
//      文件，路径经环境变量 FUSHI_READER_SELECTION_JS 传入；没传就只测 ①。
//
// 运行：node fushi/test/lookup/phrase_lookup_whitespace_bridge_bug1773_test.js

const assert = require('node:assert');
const fs = require('node:fs');
const path = require('node:path');
const vm = require('node:vm');

// ---- 最小 fake DOM -------------------------------------------------------

function makeText(content, parent) {
  return { nodeType: 3, textContent: content, nodeValue: content, parentElement: parent };
}

function selectorTags(selector) {
  return selector
    .split(',')
    .map((s) => s.trim().toLowerCase())
    .filter((s) => s.length > 0 && !s.startsWith('.') && !s.startsWith('['));
}

function makeElement(tagName) {
  const el = {
    nodeType: 1,
    tagName,
    className: '',
    parentElement: null,
    childNodes: [],
    get textContent() {
      return this.childNodes.map((c) => c.textContent).join('');
    },
    closest(selector) {
      const tags = selectorTags(selector);
      let node = this;
      while (node && node.nodeType === 1) {
        if (tags.includes((node.tagName || '').toLowerCase())) return node;
        node = node.parentElement;
      }
      return null;
    },
  };
  return el;
}

function makeRange() {
  return {
    startContainer: null,
    startOffset: 0,
    endContainer: null,
    endOffset: 0,
    setStart(node, offset) {
      this.startContainer = node;
      this.startOffset = offset;
    },
    setEnd(node, offset) {
      this.endContainer = node;
      this.endOffset = offset;
    },
    collapse() {},
    getClientRects() {
      return [{ left: 0, right: 10, top: 0, bottom: 10, x: 0, y: 0, width: 10, height: 10 }];
    },
    getBoundingClientRect() {
      return { left: 0, right: 10, top: 0, bottom: 10, x: 0, y: 0, width: 10, height: 10 };
    },
  };
}

// TreeWalker：尊重 acceptNode 的 FILTER_REJECT（阅读器版据此跳过纯空白节点）。
function makeTreeWalker(root, filter) {
  const out = [];
  (function walk(node) {
    for (const child of node.childNodes || []) {
      if (child.nodeType === 3) {
        if (!filter || filter.acceptNode(child) === 1) out.push(child);
      } else if (child.nodeType === 1) {
        walk(child);
      }
    }
  })(root);
  return {
    currentNode: root,
    nextNode() {
      const from = out.indexOf(this.currentNode);
      const next = out[from + 1] || null;
      if (next) this.currentNode = next;
      return next;
    },
  };
}

/// 建一个 <p>，其中每段文字是一个 <span> 里的文本节点（segments 长度 > 1 时用来
/// 覆盖「跨文本节点续扫」）。返回 { sandbox, textNodes }。
function buildContext(src, segments) {
  const body = makeElement('body');
  const p = makeElement('p');
  p.parentElement = body;
  body.childNodes = [p];

  const textNodes = [];
  for (const seg of segments) {
    const span = makeElement('span');
    span.parentElement = p;
    const t = makeText(seg, span);
    span.childNodes = [t];
    p.childNodes.push(span);
    textNodes.push(t);
  }

  const document = {
    body,
    createRange: makeRange,
    createTreeWalker: (root, _whatToShow, filter) => makeTreeWalker(root, filter),
    caretPositionFromPoint: () => null,
    caretRangeFromPoint: () => null,
    elementFromPoint: () => null,
  };

  const window = {
    // 所有 span 都是行内盒 → crossesRenderBoundary 判「同一行连排」，跨节点续扫放行
    // （popup 版的 BUG-1645 路径；阅读器版没有这个函数，行为一致）。
    getComputedStyle: () => ({ display: 'inline', content: 'none' }),
    getSelection: () => ({ removeAllRanges() {} }),
    flutter_inappwebview: { callHandler: () => {} },
    scanNonJapaneseText: true,
  };
  const sandbox = {
    window,
    document,
    Node: { ELEMENT_NODE: 1, TEXT_NODE: 3 },
    NodeFilter: { SHOW_TEXT: 4, FILTER_ACCEPT: 1, FILTER_REJECT: 2 },
    CSS: undefined,
    console,
  };
  sandbox.globalThis = sandbox;
  vm.createContext(sandbox);
  vm.runInContext(src, sandbox);

  // 阅读器版的收尾走 fireTextSelected → buildSelectionPayload（依赖整套
  // window.fushiReader 归一化偏移）。本测试只关心扫描产出的查询串，stub 掉收尾。
  sandbox.window.fushiSelection.fireTextSelected = function () {
    return this.selection ? this.selection.text : null;
  };
  return { sandbox, textNodes };
}

/// 跑一次取词，返回喂给引擎的查询串（没选到返回 null）。
function scan(src, segments, nodeIndex, offset, maxLength) {
  const { sandbox, textNodes } = buildContext(src, segments);
  const sel = sandbox.window.fushiSelection;
  sel.selection = null;
  sel.selectFromPosition(textNodes[nodeIndex], offset, maxLength === undefined ? 24 : maxLength);
  return sel.selection ? sel.selection.text : null;
}

// ---- 断言 ----------------------------------------------------------------

function runSuite(label, src) {
  // ① 根因回归：点 "listen" 的任意字母，查询串都必须带上后面的 " to"。
  //    修复前这里是 "listen"，短语词条永远匹配不到。
  for (const offset of [2, 4, 7]) {
    assert.strictEqual(
      scan(src, ['I listen to music.'], 0, offset),
      'listen to music',
      `[${label}] 点 listen 的第 ${offset} 位必须扫出跨空格的短语查询串（'.' 处终止）`,
    );
  }

  // ② 起点仍回退到词首（TODO-916 症状③向后兼容）。
  assert.strictEqual(
    scan(src, ['say hello world'], 0, 6),
    'hello world',
    `[${label}] 点单词中间字母仍从词首起扫`,
  );

  // ③ 连续空白终止——空格只是词间连接符，不是「吞掉排版空白」的许可证。
  assert.strictEqual(scan(src, ['a  b'], 0, 0), 'a', `[${label}] 连续空白必须终止扫描`);

  // ④ 节点末尾的空白终止（后面没有可扫字符）。
  assert.strictEqual(scan(src, ['listen '], 0, 0), 'listen', `[${label}] 节点末尾空白必须终止`);

  // ⑤ 空白后接标点终止。
  assert.strictEqual(scan(src, ['listen ,to'], 0, 0), 'listen', `[${label}] 空白后接标点必须终止`);

  // ⑥ 本节点**开头**的空白不桥接 → 跨节点不会把块间空白粘成一个词。
  assert.strictEqual(
    scan(src, ['listen', ' to'], 0, 0),
    'listen',
    `[${label}] 跨文本节点时新节点开头的空白不得桥接`,
  );

  // ⑦ 行内标签劈开的单词仍照旧续扫（BUG-1645 向后兼容，非空白续接不受影响）。
  assert.strictEqual(
    scan(src, ['lis', 'ten'], 0, 0),
    'listen',
    `[${label}] 行内标签劈开的单词必须继续粘`,
  );

  // ⑧ 日文逐字扫描不受影响。
  assert.strictEqual(
    scan(src, ['素晴らしい世界'], 0, 0),
    '素晴らしい世界',
    `[${label}] 日文扫描行为不变`,
  );

  // ⑨ maxLength 仍然是硬上限。
  assert.strictEqual(
    scan(src, ['I listen to music.'], 0, 2, 8),
    'listen t',
    `[${label}] maxLength 必须仍然截断`,
  );
}

function run() {
  const popupSrc = fs.readFileSync(
    path.resolve(__dirname, '..', '..', 'assets', 'popup', 'selection.js'),
    'utf8',
  );
  runSuite('popup/extension', popupSrc);

  const readerJsPath = process.env.FUSHI_READER_SELECTION_JS;
  if (readerJsPath && fs.existsSync(readerJsPath)) {
    runSuite('reader', fs.readFileSync(readerJsPath, 'utf8'));
  } else {
    console.log('reader selection script not provided; skipped that suite');
  }

  console.log('all assertions passed');
}

run();

File: fushi/test/lookup/phrase_lookup_whitespace_bridge_bug1773_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:fushi/src/reader/reader_selection_scripts.dart';

/// BUG-1773：英文正文/字幕里点 "listen" 只查到 listen，`listen to` 这类**空格分词
/// 短语**的词条永远匹配不到；而点它前面的空格反而能查出短语。
///
/// 根因：取词的前向扫描把空白当**终点**。`isScanBoundary` 把三件事混成一坨——
/// 空白、句读标点、「只扫日文」门控——而前向扫描直接拿它当 break 条件，于是喂给
/// 引擎的查询串被截在第一个空格前。C++ `scan_candidates` 本来就按空格分词生成
/// `listen to music` / `listen to` / `listen` 三级候选（明确禁止在单词中间切，见
/// native/fushidicts/fushidicts_src/scan/word_scan.cpp），拿不到空格就等于把短语
/// 整类排除在匹配之外。
///
/// 修复：拆成两个语义不同的谓词——
/// - `isScanBoundary`（**词边界**：点击命中判定 + 词首回退用，含空白，语义不变）
/// - `isScanStop`（**扫描终点**：标点 / 门控，不含空白）
/// 空白能否跨过去由桥接规则单独决定：只在同一文本节点内部跨、且只跨一个（左边必须
/// 已有本节点扫入的内容，右边必须紧跟一个可扫字符）。
///
/// 两层守护：
/// ① 行为级——用 node 真执行两份实现（浮窗/扩展的 selection.js + 阅读器注入脚本
///    `ReaderSelectionScripts.source()`）的 `selectFromPosition`，断言跨空格短语、
///    以及连续空白/末尾空白/空白接标点/跨节点开头空白四条终止规则。无 node 时 skip。
/// ② 源码级——两份实现的前向扫描都必须用 `isScanStop` + 空白桥接，不得退回
///    `isScanBoundary`（那正是 BUG-1773 的写法）。
void main() {
  const List<String> selectionCopies = <String>[
    'assets/popup/selection.js',
    'assets/browser_extension/vendor/selection.js',
    '../tools/browser-extension/vendor/selection.js',
  ];

  test(
    'BUG-1773: forward scan bridges a single space so phrases stay reachable '
    '(executes both selection engines via node)',
    () async {
      final String? nodeExe = _resolveNode();
      if (nodeExe == null) {
        markTestSkipped(
            'node not found on PATH; skipping JS behavior execution');
        return;
      }

      final File jsTest = File(
        'test/lookup/phrase_lookup_whitespace_bridge_bug1773_test.js',
      );
      expect(jsTest.existsSync(), isTrue,
          reason: 'behavior harness ${jsTest.path} must exist');

      // 阅读器注入脚本活在 Dart 的 raw string 里，node 读不到文件。把**真值**
      // （ReaderSelectionScripts.source()，与真机注入的是同一份字符串）落到临时
      // 文件再交给 harness，避免测试对着一份手抄副本自娱自乐。
      final Directory tmp =
          Directory.systemTemp.createTempSync('fushi_bug1773_');
      final File readerJs = File('${tmp.path}/reader_selection.js');
      readerJs.writeAsStringSync(ReaderSelectionScripts.source());

      try {
        final ProcessResult result = await Process.run(
          nodeExe,
          <String>[jsTest.path],
          workingDirectory: Directory.current.path,
          environment: <String, String>{
            'FUSHI_READER_SELECTION_JS': readerJs.path,
          },
        );

        expect(
          result.exitCode,
          0,
          reason: 'BUG-1773 whitespace-bridge behavior test failed.\n'
              'stdout:\n${result.stdout}\nstderr:\n${result.stderr}',
        );
        final String stdout = result.stdout.toString();
        expect(stdout, contains('all assertions passed'),
            reason: 'behavior harness must reach its success marker');
        // 阅读器那一套必须真的跑到，否则这个测试只守住了一半实现。
        expect(stdout, isNot(contains('reader selection script not provided')),
            reason: '阅读器注入脚本必须被 harness 真执行（真值来自 source()）');
      } finally {
        tmp.deleteSync(recursive: true);
      }
    },
  );

  test('forward scan uses isScanStop + whitespace bridge, not isScanBoundary',
      () {
    final Map<String, String> sources = <String, String>{
      for (final String path in selectionCopies)
        path: File(path).readAsStringSync(),
      'reader_selection_scripts.dart (source())':
          ReaderSelectionScripts.source(),
    };

    // 断言一律锚在**可执行代码片段**上，不能是裸的标识符名——两份实现的注释里都
    // 写着 isScanStop / isScanBoundary / `scanOffset === start`，裸 contains 会被
    // 自己的注释假阳性满足，变异测试实测过（把桥接条件换成 `!text` 时裸断言照样绿）。
    sources.forEach((String label, String src) {
      // 前向扫描的 break 条件必须是 isScanStop。BUG-1773 的写法是 isScanBoundary，
      // 它把空白也当终点 → 短语被截断。
      expect(src.contains('if (this.isScanStop(char)) break;'), isTrue,
          reason: '[$label] 前向扫描必须以 isScanStop 为终点');
      expect(src.contains('if (this.isScanBoundary(char)) break;'), isFalse,
          reason: '[$label] 前向扫描不得退回 isScanBoundary（BUG-1773 的根因写法）');

      // 桥接规则的三条终止条件必须齐：本节点开头 / 后无字符 / 后接空白或终点。
      // 两份实现缩进不同，故锚在去缩进的完整条件表达式上。
      expect(
          src.contains('if (scanOffset === start || nextChar === undefined ||'),
          isTrue,
          reason: '[$label] 空白桥接必须要求左边已有本节点扫入的内容（跨节点不粘空白）'
              '、且节点末尾的空白终止扫描');
      expect(
          src.contains('this.isScanWhitespace(nextChar) || '
              'this.isScanStop(nextChar)) break;'),
          isTrue,
          reason: '[$label] 连续空白 / 空白后接终点必须终止扫描');

      // 词首回退仍必须用词边界语义（含空白），否则回退会越过空格吃进上一个词。
      expect(src.contains('!this.isScanBoundary(hitContent[startOffset - 1])'),
          isTrue,
          reason: '[$label] 词首回退必须继续用 isScanBoundary（含空白）');
    });
  });
}

/// Resolve a usable `node` executable, returning null when none is on PATH.
String? _resolveNode() {
  final List<String> candidates =
      Platform.isWindows ? <String>['node.exe', 'node'] : <String>['node'];
  for (final String name in candidates) {
    try {
      final ProcessResult probe = Process.runSync(name, <String>['--version']);
      if (probe.exitCode == 0) {
        return name;
      }
    } on ProcessException {
      // Not found; try next candidate.
    }
  }
  return null;
}
